### PR TITLE
post: cli 1.6

### PR DESCRIPTION
**Pull request by cexbrayat**
WIP

- [x] service worker
- [x] universal
- [x] app shell

### CODE REVIEW

**Comment by cexbrayat** [comment]
Haha, is the part about service worker not clear? (the rest is still a very early draft).

**Comment by cexbrayat** [comment]
So the "application shell" feature is still buggy, but I managed to test it on a simple app.
The blog post is scheduled for Tuesday.

@clacote @jnizet As usual, would be great to have your reviews

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+This is now fixed: you just have to run `ng g universal server-app`
+and Angular CLI automatically sets things up for you:
+
+- creates a new module `app.server.module.ts`, which uses `ServerModule` instead of `BrowserModule`;
+- creates a main file `main.server.ts`;
+- creates a TypeScript config `tsconfig.server.ts`
+- adds a dependency to your package.json `@angular/platform-server`;
+- updates `.angular-cli.json` to add a new "app", called "server-app" in my example, in your project;
+- updates your main module to use the brand new (Angular 5) `BrowserModule.withServerTransition()` which ensures the client app has a smooth transition when it boots;
```
Comment (clacote) [inline _posts/2017-12-12-angular-cli-1.6.md]: Don't know if it's important, but I don't understand what "a smooth transition" means in this case.

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+- wraps the bootstrap call in a `DOMContentLoaded` event handler.
+
+You can customize pretty much everything:
+
+- the app's `name`, by default the one you gave in the command line
```
Comment (clacote) [inline _posts/2017-12-12-angular-cli-1.6.md]: I'm not sure, but I thought the gerundive form `'s` has only to be applied to people, not to things.

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+- `index`, the name of the index file, by default `index.html`
+- `main`, the name of the entry point, by default `main.server.ts`
+- `test`, the entry point for the tests, by default `test.ts`
+- `tsConfigFileName`, the name of the TS config file, by default `tsconfig.server.json`
+- `testTsconfigFileName`, the name of the TS config file for the tests, by default `tsconfig.spec.json`
+- `appDir`, the name of the app's directory, by default `app`
+- `rootModuleFileName`, the name of the module file, by default `app.server.module.ts`
+- `rootModuleClassName`, the name of the module class, by default `AppServerModule`
+
+Note that you'll have to add the deprecated `@angular/http` package even you don't use it,
```
Comment (clacote) [inline _posts/2017-12-12-angular-cli-1.6.md]: `even you don't use it` -> `even if you don't use it`, no?

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+because `@angular/platform-server` still depends on it.
+
+Now you should be able to build the application!
+
+    ng build --prod --app=server-app
+
+And you'll get the bundles in `dist-server/`. Than it's up to you to set up a server to use these.
+Or you can continue reading, and discover another new feature of the CLI ;)
```
Comment (clacote) [inline _posts/2017-12-12-angular-cli-1.6.md]: No final punctuation after emoji.

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+The command here uses the `serverApp` I created above.
+If you did not create it, the CLI will generate it for you ;)
+
+The command will:
+
+- update `.angular-cli.json` to add the `appShell` key with its configuration `"appShell": { "app": "server-app", "route": "shell" }`;
+- add an `AppShellComponent` to your application;
+- updates the `app.server.module.ts` to add a route with the path specified pointing to the `AppShellComponent`.
+
+Now when you'll build the application with `ng build --prod` the `dist` and `dist-server` diretories will be generated,
```
Comment (clacote) [inline _posts/2017-12-12-angular-cli-1.6.md]: diretories -> directories

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+[Angular CLI 1.6.0](https://github.com/angular/angular-cli/releases/tag/v1.6.0) is out with some nice new features!
+
+If you want to upgrade to 1.6.0 without pain (or to any other version, BTW), I have created a Github project to help: [angular-cli-diff](https://github.com/cexbrayat/angular-cli-diff). Choose the version you're currently using (1.2.1 for example), and the target version (1.6.0 for example), and it gives you a diff of all files created by the CLI: [angular-cli-diff/compare/1.2.1…1.6.0](https://github.com/cexbrayat/angular-cli-diff/compare/1.2.1...1.6.0). You have no excuse for staying behind anymore!
+
+Let's see what new features we have!
+
+## Service Worker
+
+A new flag has been added to the `new` command and allows to generate an application with service workers already enabled and configured.
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: `new` -> `ng new`.

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+      ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production })
+    ],
+
+This will activate the `ServiceWorkerModule` if you are in "prod" mode,
+and register the service worker generated by the CLI based on the previous config.
+As the worker is only enabled in "prod" mode, you won't see a difference using `ng serve`.
+It's only when building your application with `ng build --prod` that you'll see a few new files in your `dist` directory.
+
+As this module is in `@angular/service-worker`, it will also add this package to your dependencies.
+It will also turn on the `serviceWorker` option in your `.angular-cli.json`.
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: It will also -> And it will

to avoid the repetition

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+Here I cache the calls to `/api/races` for 30 minutes,
+meaning all requests to this endpoint in the next half hour will return the same result as the first one,
+even if offline.
+
+You can also add external resources like fonts to the configuration to also cache them.
+
+To give it a try, build your app with `ng build --prod`,
+then serve your `dist` directory with a static HTTP server,
+like `http-server`. Open your application with Chrome,
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: Why not use `ng serve --prod`?

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+and go the Developer Tools in the Application tab.
+You should see the service worker registered in the Service Worker section!
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: You should see an entry in the *Service Worker* section. And you can check which parts are being cached in the *Cache Storage* section.

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+              () => this.swUpdate.activateUpdate().then(() => document.location.reload()), // load the update
+              () => {} // do nothing if the user wants to refresh later
+            )
+          );
+        }
+      }
+    }
+
+Here I'm subscribing to the `available` observable exposed by `SwUpdate`.
+As soon as a new version of the application will be available,
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: is available

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+
+## Universal
+
+One of the main features of this release is the ability to add server side rendering capacity to your application
+with a simple command.
+Server-side rendering (called "universal" in Angular) is possible since quite some time,
+but the CLI had no automated way to do it.
+
+This is now fixed: you just have to run `ng g universal server-app`
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: ng generate

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+- creates a TypeScript config `tsconfig.server.ts`
+- adds a dependency to your package.json `@angular/platform-server`;
+- updates `.angular-cli.json` to add a new "app", called "server-app" in my example, in your project;
+- updates your main module to use the brand new (Angular 5) `BrowserModule.withServerTransition()` which ensures the client app has a smooth transition when it boots;
+- wraps the bootstrap call in a `DOMContentLoaded` event handler.
+
+You can customize pretty much everything:
+
+- the app's `name`, by default the one you gave in the command line
+- the `clientApp`'s name, by default `0`, as you usually have only one app in your CLI project
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: I don't understand what this clientApp is. Isn't there an already existing client app (ponyracer, for example)?

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+- `outDir`, the output directory of the build, by default `dist-server/`
+- `root`, the source directory, by default `src`
+- `index`, the name of the index file, by default `index.html`
+- `main`, the name of the entry point, by default `main.server.ts`
+- `test`, the entry point for the tests, by default `test.ts`
+- `tsConfigFileName`, the name of the TS config file, by default `tsconfig.server.json`
+- `testTsconfigFileName`, the name of the TS config file for the tests, by default `tsconfig.spec.json`
+- `appDir`, the name of the app's directory, by default `app`
+- `rootModuleFileName`, the name of the module file, by default `app.server.module.ts`
+- `rootModuleClassName`, the name of the module class, by default `AppServerModule`
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: Not sure if it's really useful to be so exhaustive. It's a blog post, not a reference documentation.

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+Note that you'll have to add the deprecated `@angular/http` package even you don't use it,
+because `@angular/platform-server` still depends on it.
+
+Now you should be able to build the application!
+
+    ng build --prod --app=server-app
+
+And you'll get the bundles in `dist-server/`. Than it's up to you to set up a server to use these.
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: What kind of server should that be? A NodeJS server? Something else? What exactly is being generated? The big picture, explaining what the server app is all about, and how it relates and combines with the client app, is missing.

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+Or you can continue reading, and discover another new feature of the CLI ;)
+You won't be able to use `ng serve` with this app though (not yet, but I think it's coming!).
+
+
+## Application Shell
+
+The other new feature is the Application Shell.
+A new schematic has been added allowing to generate a "shell" - an `index.html` containing a static rendering of your application. This uses the `universal` part I mentioned above for the server-side rendering.
+You can even specify the route path you want to serve this application shell!
```
Comment (jnizet) [inline _posts/2017-12-12-angular-cli-1.6.md]: from, or in, this application shell, no?

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 8 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: If I remember correctly, it does not activate the the build of the service worker files.

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 12 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: Yes, that's why by default you don't really care. But you could have several applications in the same CLI project, and this allows you to pick the one you want.

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 1 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: It allows the client application to know what has been created by the server side application and remove the proper elements, reuse some, create new ones etc...

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 13 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: I removed the less important options

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 14 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: I added a line with a link to the official documentation.

**Comment by cexbrayat** [comment]
I think we're good with the last commit. If you don't spot anything left, OK to merge tomorrow morning, and tweet about it, smthg like:

```
Angular CLI 1.6 is out! ✨

Check out our blog post to know what's new: service worker, server-side rendering, application shell!
```

@clacote can you handle it? I'll be on the road a fair share of the day

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+            )
+          );
+        }
+      }
+    }
+
+Here I'm subscribing to the `available` observable exposed by `SwUpdate`.
+As soon as a new version of the application is available,
+the observable will emit an event.
+In the example above, I chose to open a modal with `ng-bootstrap` to ask the user to reload the application.
```
Comment (acrepet) [inline _posts/2017-12-12-angular-cli-1.6.md]: inclusive writing! ;-)

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+If he/she chooses to do so, we load the new version with `activateUpdate()` and reload the page. If not, we do nothing.
```
Comment (acrepet) [inline _posts/2017-12-12-angular-cli-1.6.md]: inclusive writing! ;-)

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+This is now fixed: you just have to run `ng generate universal server-app`
+and Angular CLI automatically sets things up for you:
+
+- creates a new module `app.server.module.ts`, which uses `ServerModule` instead of `BrowserModule`;
+- creates a main file `main.server.ts`;
+- creates a TypeScript config `tsconfig.server.ts`
+- adds a dependency to your package.json `@angular/platform-server`;
+- updates `.angular-cli.json` to add a new "app", called "server-app" in my example, in your project;
+- updates your main module to use the brand new (Angular 5) `BrowserModule.withServerTransition()` which ensures the client app has a smooth transition when it boots;
+- wraps the bootstrap call in a `DOMContentLoaded` event handler.
```
Comment (acrepet) [inline _posts/2017-12-12-angular-cli-1.6.md]: which of plural /singlar form is the most appropriate : bootstrap call or bootstrap calls ?

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+    ng g app-shell appShell --universal-app=server-app
+
+The command here uses the `serverApp` I created above.
+If you did not create it, the CLI will generate it for you ;)
+
+The command will:
+
+- update `.angular-cli.json` to add the `appShell` key with its configuration `"appShell": { "app": "server-app", "route": "shell" }`;
+- add an `AppShellComponent` to your application;
+- updates the `app.server.module.ts` to add a route with the path specified pointing to the `AppShellComponent`.
```
Comment (acrepet) [inline _posts/2017-12-12-angular-cli-1.6.md]: updates -> update

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+Now when you'll build the application with `ng build --prod` the `dist` and `dist-server` directories will be generated,
+and the `index.html` file in `dist` will contain the prerendered `AppShellComponent`.
+This means you'll have a meaningful content served right away when deploying to production
+(as you can of course put whatever you want in the shell component).
+When the application is done starting, it will be displayed as usual.
```
Comment (acrepet) [inline _posts/2017-12-12-angular-cli-1.6.md]: what do you want to mean with "done starting", "starting" is enought, isn't?

File: _posts/2017-12-12-angular-cli-1.6.md [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+
+The three features can be combined nicely: you can add a universal/server-side rendering capability to your application,
```
Comment (acrepet) [inline _posts/2017-12-12-angular-cli-1.6.md]: universal/server-side -> `universal/server-side`

**Comment by acrepet** [comment]
so sorry for my late (and probably not very consistent) review ;-(

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 17 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: I try as often as possible (but sometimes forget) :)

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 18 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: Singular should be OK here as there is only one bootstrap call in most applications.

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 20 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: I mean when it has completed the bootstrap phase. Maybe 'started' would be better, but I think 'done starting' is OK

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 21 above)
Comment (cexbrayat) [reply _posts/2017-12-12-angular-cli-1.6.md]: We usually use this for package names and even if it looks like one it's not :) I used the / to show the two common names : universal is the official name, but most frameworks are talking about server-side rendering.

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 18 above)
Comment (acrepet) [reply _posts/2017-12-12-angular-cli-1.6.md]: ok!

File: _posts/2017-12-12-angular-cli-1.6.md (hunk 20 above)
Comment (acrepet) [reply _posts/2017-12-12-angular-cli-1.6.md]: ok!

**Comment by cexbrayat** [comment]
Feedback (even late) is always welcome :) Thx for reviewing it, my last commit should fix the typos the spotted 👍

**Comment by clacote** [comment]
@cexbrayat will merge and tweet

**Comment by clacote** [comment]
https://twitter.com/NinjaSquad/status/940493948929413120

**Pull request merged**